### ISSUE TITLE
Hanging Firestore queries on version 10.4.0

### ISSUE THREAD

**Issue by thomasdao** [firebase/firebase-js-sdk#7652]
### Operating System

Both Mac and Windows

### Browser Version

Electron 24.8.3

### Firebase SDK Version

10.4.0

### Firebase SDK Product:

Auth, Firestore, Functions, Storage

### Describe your project's tooling

Electron 24.8.3

### Describe the problem

Hi, since we updated Firebase to 10.4.0, we have received multiple reports from our users that they can't retrieve the data from Firestore server. We then downgraded to 10.3.1 and these users can retrieve data normally. From the release note, we believe this is related:

```
Fixed an issue where Cloud Firestore was incorrectly using XHR instead of fetch for streaming response.

Updated dependency grpc-js from version ~1.8.17 to ~1.9.0.
```

I also noticed that on 10.4.0, our query becomes a lot slower and frequently cannot complete. Switch back to 10.3.1 and the query becomes a lot faster.

### Steps and code to reproduce issue

This issue only happens for 10.4.0.

**Comment by thomasdao** [comment]
When I set the log level to debug, there're a lot of messages in the console as below, but I'm not sure if it's related:

```
[2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 10
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 42
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 44
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 46
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 48
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 50
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 52
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): WatchChangeAggregator Detected inactive target 54
index.esm2017.js:78 [2023-09-27T04:11:34.774Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Get last remote snapshot version
index.esm2017.js:78 [2023-09-27T04:11:34.775Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Apply remote event
index.esm2017.js:78 [2023-09-27T04:11:34.777Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: notifyLocalViewChanges
index.esm2017.js:78 [2023-09-27T04:11:34.777Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Release target
index.esm2017.js:78 [2023-09-27T04:11:34.780Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 sending: {"database":"projects/xxxxxx/databases/(default)","removeTarget":30}
index.esm2017.js:78 [2023-09-27T04:11:34.785Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Release target
index.esm2017.js:78 [2023-09-27T04:11:34.788Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 sending: {"database":"projects/xxxxxx/databases/(default)","removeTarget":32}
index.esm2017.js:78 [2023-09-27T04:11:34.793Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Release target
index.esm2017.js:78 [2023-09-27T04:11:34.796Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 sending: {"database":"projects/xxxxxx/databases/(default)","removeTarget":34}
index.esm2017.js:78 [2023-09-27T04:11:34.802Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Release target
index.esm2017.js:78 [2023-09-27T04:11:34.805Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 sending: {"database":"projects/xxxxxx/databases/(default)","removeTarget":36}
index.esm2017.js:78 [2023-09-27T04:11:34.809Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Release target
index.esm2017.js:78 [2023-09-27T04:11:34.812Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 sending: {"database":"projects/xxxxxx/databases/(default)","removeTarget":38}
index.esm2017.js:78 [2023-09-27T04:11:34.817Z]  @firebase/firestore: Firestore (10.4.0): MemoryPersistence Starting transaction: Release target
index.esm2017.js:78 [2023-09-27T04:11:34.820Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 sending: {"database":"projects/xxxxxx/databases/(default)","removeTarget":40}
index.esm2017.js:78 [2023-09-27T04:11:34.881Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[42]}}
index.esm2017.js:78 [2023-09-27T04:11:34.882Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[44]}}
index.esm2017.js:78 [2023-09-27T04:11:34.882Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[46]}}
index.esm2017.js:78 [2023-09-27T04:11:34.882Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[48]}}
index.esm2017.js:78 [2023-09-27T04:11:34.882Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[50]}}
index.esm2017.js:78 [2023-09-27T04:11:34.882Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[52]}}
index.esm2017.js:78 [2023-09-27T04:11:34.882Z]  @firebase/firestore: Firestore (10.4.0): WebChannelConnection RPC 'Listen' stream 0x3e730da7 received: {"targetChange":{"targetChangeType":"ADD","targetIds":[54]}}
```

**Comment by dconeybe** [comment]
Drive-by comment: I'm looking into another issue that has those strange "Detected inactive target" log messages: https://github.com/firebase/firebase-js-sdk/issues/7641. I wonder if that is related to this one.

@thomasdao: Can you provide a longer debug log than the one in your comment above? I'm quite interested to see what log messages appeared _before_ the "Detected inactive target" log messaged.

**Comment by thomasdao** [comment]
Hi @dconeybe, I can send you the whole debug log, however since it contains some private data, how do I send it to you? Thanks!

**Comment by dconeybe** [comment]
> Hi @dconeybe, I can send you the whole debug log, however since it contains some private data, how do I send it to you?

The best way would be to go to https://firebase.google.com/support, select "Bugs and Features", select "Bug Report", and open a bug. You can attach the log to that bug. Note that there is a disclaimer about PII at the bottom of that page so make sure that you're okay with that before uploading your logs. Make sure to include the URL of this GitHub issue in the issue's description as well as my username. The ticket should get routed to me, but it could take a day or two. I'm especially interested in the "addTarget" and "removeTarget" messages, in addition to the logs messages that you've already posted.

Alternately, feel free to redact the sensitive information from the log before posting it, such as the database ID, document names, and document contents, and post the logs publicly.

Could you tell me, are the documents that seem to be causing this problem "large", in that they have a lot of fields?

**Comment by thomasdao** [comment]
Hi @dconeybe, I added the log below, some data has been redacted with `<<< Data received >>>`. For easier comparison, I attach the log for both 10.3.1 and 10.4.0.

For 10.3.1, the query completed within a few seconds, but for 10.4.0, the query still stucks for more than 10 minutes at the time I took the log.

> Could you tell me, are the documents that seem to be causing this problem "large", in that they have a lot of fields?

My document contains around 10 fields, usually less than 100KB. I used pagination to retrieve data, each page contains only 100 documents.

Log for 10.4.0: https://drive.google.com/file/d/1yHTwXPLw9FWi7M5ZWV4Ije-cNw8fnbVH/view?usp=sharing

Log for 10.3.1: https://drive.google.com/file/d/1Dq7ypEXYp1BLEwIXGOH_pYdHbMIBq3TE/view?usp=sharing

**Comment by dconeybe** [comment]
Update: After comparing the logs for this issue with the logs from the potentially-related issue https://github.com/firebase/firebase-js-sdk/issues/7641, it appears that they are distinct issues with the same presenting problem (i.e. the callback registered with `onSnapshot()` is not being called).

Namely, in #7641 the root cause appears to be the server sending multiple `targetChangeType=ADD` messages for a single target, causing the client's internal state to become invalid and erroneously discard the document data that it is sent. In comparison, the logs for this issue do _not_ show multiple `targetChangeType=ADD` messages for a single target (i.e. they show exactly 1 such message, as expected), leading me to conclude that it has a different root cause and needs to be investigated independently.

I'm assigning @MarkDuckworth who is on support rotation at the moment.

**Comment by MarkDuckworth** [comment]
@thomasdao, This would be easiest to diagnose if we can reproduce it. Can you provide some example code that shows us what your application is doing, or preferably, can you provide a [minimal reproducible example electron app](https://stackoverflow.com/help/minimal-reproducible-example)?

Also, since an Electron app has both node.js and chromium processes, we inferred from the logs that the failing code is running in chromium/browser window. Can you confirm that your affected code is only running in the browser window? If so, this rules out the grpc-js version change.

**Comment by thomasdao** [comment]
@MarkDuckworth I setup Firebase on Web Worker of a browser window, however the browser window was setup with extra options `nodeIntegration: true` and`nodeIntegrationInWorker: true`. In my code, I retrieve documents from server as below. This code works fine for all Firebase version except 10.4.0.

```js
async function retrievePaginatedData (options) {
  let qry = getQuery(options)
  let snapshots = await getDocsFromServer(qry)
  const { namespace, batchSize } = options

  postBatchData(namespace, snapshots)
  if (batchSize) {
    let docs = snapshots.docs
    while (docs.length === batchSize) {
      const lastSnapshot = docs[docs.length-1]
      const newOpts = Object.assign({}, options, {
        lastSnapshot
      })
      qry = getQuery(newOpts)

      if (self.gc) {
        self.gc()
      }

      snapshots = await getDocsFromServer(qry)

      // Update result array
      postBatchData(namespace, snapshots)

      // Update docs length
      docs = snapshots.docs
    }
  }

  if (self.gc) {
    self.gc()
  }
}

function getQuery (options) {
  const user = getAuth().currentUser
  if (!user) return null

  const {
    namespace, 
    queryTime = null,
    lastQueryTime = null,
    batchSize = null, 
    lastSnapshot = null, 
    shouldQueryWithSyncedAt = false
  } = options

  const userId = user.uid

  const firestore = getFirestore()

  const colRef = collection(firestore, namespace)
  const args = [colRef]

  if (queryTime) {
    args.push(where('timestamp', '<', new Date(time)))
  }

  if (lastQueryTime) {
    args.push(where('timestamp', '>', new Date(time)))
  }

  if (batchSize) {
    args.push(orderBy('timestamp', 'desc'))
    args.push(orderBy('id'))

    if (lastSnapshot) {
      args.push(startAfter(lastSnapshot))
    }

    args.push(limit(batchSize))
  }

  return query(...args)
}

// Post data to host window
function postBatchData (namespace, querySnapshot) {
  const queryData = []

  querySnapshot.forEach(docData => {
    queryData.push(docData.data())
  })

  promWorker.postMessage({
    type: 'ON_SNAPSHOT',
    content: {
      namespace: namespace,
      result: queryData
    }
  })

  return queryData
}

```

**Comment by thomasdao** [comment]
@wu-hui hi, I've created a minimal sample project on GitHub and have invited you as collaborator. Please let me know if you can reproduce the issue on your side, thank you!

**Comment by dconeybe** [comment]
@MarkDuckworth are you able to take a look?

**Comment by MarkDuckworth** [comment]
This may be the same issue as https://github.com/firebase/firebase-js-sdk/issues/7641. I was able to use the repro from @thomasdao, and I saw duplicate target adds for a single target ID. This reproduction is very helpful and should help us resolve the issue faster.

**Comment by thomasdao** [comment]
Hi @dconeybe, I know about `firestore-lite`, however in my project I need to use `firestore` because I need to listen to data changes from the server. I hope @MarkDuckworth can find out the issue :)

**Comment by MarkDuckworth** [comment]
@thomasdao, the data in the collection used by your test app, is that test data that can be copied into our database? I'm having trouble reproducing using your reproduction app against my own data in my db.

**Comment by thomasdao** [comment]
@MarkDuckworth yes you can clone to your database, please delete them when you're done, thank you!

**Comment by MarkDuckworth** [comment]
@thomasdao, we released v10.5.2, which rolls back the change in v10.4.0 that was responsible for hanging queries. I verified the fix on two reproductions, including yours. Thanks for your help figuring this out!
